crates/hir_ty/src/infer/closure.rs

```diff
@@ -42,46 +42,38 @@ impl InferenceContext<'_> {
 
         let fn_traits: SmallVec<[ChalkTraitId; 3]> =
             utils::fn_traits(self.db.upcast(), self.owner.module(self.db.upcast()).krate())
-                .map(|tid| to_chalk_trait_id(tid))
+                .map(to_chalk_trait_id)
                 .collect();
 
         let self_ty = TyKind::Error.intern(&Interner);
         let bounds = dyn_ty.bounds.clone().substitute(&Interner, &[self_ty.cast(&Interner)]);
         for bound in bounds.iter(&Interner) {
             // NOTE(skip_binders): the extracted types are rebound by the returned `FnPointer`
-            match bound.skip_binders() {
-                WhereClause::AliasEq(eq) => match &eq.alias {
-                    AliasTy::Projection(projection) => {
-                        let assoc_data = self.db.associated_ty_data(projection.associated_ty_id);
-                        if !fn_traits.contains(&assoc_data.trait_id) {
-                            return None;
-                        }
+            if let WhereClause::AliasEq(AliasEq { alias: AliasTy::Projection(projection), ty }) =
```

Comment:
Probably intentional

Reply:
For what reason? To me collapsing the matches seems better personally

Reply:
ditto: here, a complicated pattern comes *before* the expression to match.